CMS-Enterprise/sbom-harbor
# feat(multi-target-repo): add support for multi-target repositories

## Summary

Design and implement a way to extract every target (build file) from a given repository using the GitHub Client

### Added

- Git Service to handle Git Client specific operations.
- Syft Service to manage Syft actions.

### Changed

- Refactored GitHub Provider Sync service to handle searching for build files associated to a particular cataloger
- Now GitHub Provider will create an SBOM for every build target specified in the configuration

## How to test

- Start test environment
- Run `/target/debug/harbor ingest --provider github harbor-test-org --debug`
- Verify that:
  - 18 `primary` SBOMs are produced and are in MongoDB 
  - Provider produces 1 SBOM for: `https://github.com/harbor-test-org/java-repo`. Look in MongoDB for purl like:
    - pkg:**java-pom**:harbor-test-org/**java-repo**@|commit-hash|
    - In the example above, the important pieces of data are: 
      - `java-pom`: The cataloger used to create the SBOM.
      - `java-repo`: The name of the repository the SBOM was created from
      - `harbor-test-org`: is also important because it is the organization the repos are in  
  - Provider produces 11 SBOMS from the java multi repo like:
    - pkg:**java-pom**:harbor-test-org/**java-multi-module**@|commit-hash|#**lib/api-docs**
    - In the example above, also look to see that the file location paths are unique
  - Provider produces 2 SBOMs from the `ror-repo-with-gemfile` repo.  One from Ruby and one from JavaScript:
    - pkg:**ruby-gemfile**:harbor-test-org/**ror-repo-with-gemfile**@|commit-hash|
    - pkg:**javascript-package**:harbor-test-org/**ror-repo-with-gemfile**@|commit-hash|
  - Provider produces 1 SBOM from the `python-project-with-requirements-txt` repo:
    - pkg:**python-index**:harbor-test-org/**python-project-with-requirements-txt**@|commit-hash|
  - Provider produces 1 SBOM from the `javascript-repo-correct` repo:
    - pkg:**javascript-package**:harbor-test-org/**javascript-repo-correct**@|commit-hash|
  - Provider produces only 1 SBOM from `javascript-repo-node-modules-checked-in`.  This is important, because the 
    node-modules folder is checked in and populated with dependencies.  This tests the ability for the Provider to avoid
    processing build targets when they are under specified directories.  This feature is mandatory because I have 
    identified at least one case of this in the CMS repos. 
    - pkg:**javascript-package**:harbor-test-org/**javascript-repo-node-modules-checked-in**@|commit-hash|
  - Provider produces 1, `no-cataloger` SBOM in the `terraform-only` repo as there are no build targets.
    - pkg:**no-cataloger**:harbor-test-org/**terraform-only**@|commit-hash|

### Points of interest:
- Forked repos are not processed by the GitHub Provider. They must be repos that are created in a given organization to be processed. If we want to process Forks as well, we will have to refactor.
- This iteration of the GitHub provider supports: Java, Javascript, Python, Ruby and Hail Mary repos with no build target
- Currently, the configuration for the catalogers and build files resides in `sdk/core/src/services/github/service.rs`.  This config should be encoded in JSON and moved somewhere that makes sense
